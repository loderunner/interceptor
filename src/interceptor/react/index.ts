export { Provider as InterceptProvider, useIntercepts } from './intercept'
