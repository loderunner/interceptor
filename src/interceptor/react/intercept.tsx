import * as React from 'react'
import { createContext, useContext, useState } from 'react'

import * as intercept from '../intercept'

interface InterceptContextType {
  intercepts: intercept.Intercept[]
  addIntercept: (pattern: string) => void
  removeIntercept: (inter: intercept.Intercept) => void
}

const InterceptContext = createContext<InterceptContextType>({
  intercepts: [],
  addIntercept: () => {},
  removeIntercept: () => {},
})

interface Props {
  children: React.ReactNode
}

export const Provider = ({ children }: Props) => {
  const [intercepts, setIntercepts] = useState([...intercept.intercepts])

  const addIntercept = (pattern: string) => {
    intercept.addIntercept({ pattern, enabled: true })
    setIntercepts([...intercept.intercepts])
  }
  const removeIntercept = (inter: intercept.Intercept) => {
    intercept.removeIntercept(inter)
    setIntercepts([...intercept.intercepts])
  }

  return (
    <InterceptContext.Provider
      value={{ intercepts, addIntercept, removeIntercept }}
    >
      {children}
    </InterceptContext.Provider>
  )
}

export const useIntercepts = () => useContext(InterceptContext)
