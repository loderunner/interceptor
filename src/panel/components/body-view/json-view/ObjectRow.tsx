import * as React from 'react'
import { useCallback, useMemo, useState } from 'react'

import {
  Add as AddIcon,
  Clear as ClearIcon,
  Remove as RemoveIcon,
} from '@/icons'

import { isArray, isObject } from '../JSON'

import KeyView from './KeyView'
import ObjectView from './ObjectView'

import type { JSONArray, JSONObject, JSONValue } from '../JSON'

interface Props {
  k: string
  v: JSONArray | JSONObject
  keyEditable: boolean
  depth: number
  onChangeKey?: (k: string) => void
  onChangeValue?: (v: JSONValue) => void
  onDelete?: () => void
}

const ObjectRow = ({
  k,
  v,
  keyEditable,
  depth,
  onChangeKey,
  onChangeValue,
  onDelete,
}: Props) => {
  const [hovering, setHovering] = useState(false)
  const [folded, setFolded] = useState(depth >= 1)

  const [leftGuard, rightGuard] = useMemo(() => {
    if (isArray(v)) {
      return ['[', ']']
    } else if (isObject(v)) {
      return ['{', '}']
    }
    throw new Error('invalid JSON value')
  }, [v])

  const onToggleFolded = useCallback(() => {
    setFolded((f) => !f)
  }, [])

  const foldButton = useMemo(() => {
    return (
      <button
        className="mx-0.5 rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
        onClick={onToggleFolded}
      >
        {folded ? <AddIcon /> : <RemoveIcon />}
      </button>
    )
  }, [folded, onToggleFolded])

  return (
    <>
      <div
        className="mt-1 flex items-center space-x-1"
        style={{ marginLeft: `${depth}rem` }}
        onMouseEnter={() => setHovering(true)}
        onMouseLeave={() => setHovering(false)}
      >
        <button
          className={`rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 ${
            hovering ? '' : 'invisible'
          }`}
          onClick={onDelete}
        >
          <ClearIcon className="fill-white" />
        </button>
        <KeyView name={k} editable={keyEditable} onChange={onChangeKey} />
        <pre className="flex items-center space-x-0.5">
          {leftGuard}
          {foldButton}
          {rightGuard}
        </pre>
      </div>
      <ObjectView
        className={`${folded ? 'hidden' : ''}`}
        obj={v}
        depth={depth + 1}
        onChange={onChangeValue}
      />
    </>
  )
}

export default ObjectRow
