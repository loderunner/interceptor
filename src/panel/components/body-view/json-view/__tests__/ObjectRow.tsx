import { cleanup, render } from '@testing-library/react'
import * as React from 'react'
import '@testing-library/jest-dom'

import ObjectRow from '../ObjectRow'

describe('[JSON.ObjectRow]', () => {
  afterEach(() => {
    cleanup()
  })

  it('should match initial snapshot', () => {
    const { container } = render(
      <ObjectRow
        depth={0}
        k="key"
        v={JSON.parse(globalMocks.request.postData as string)}
        keyEditable
      />
    )
    expect(container).toMatchSnapshot()
  })
})
