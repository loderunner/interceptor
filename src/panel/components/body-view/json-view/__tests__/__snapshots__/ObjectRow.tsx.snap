// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[JSON.ObjectRow] should match initial snapshot 1`] = `
<div>
  <div
    class="mt-1 flex items-center space-x-1"
    style="margin-left: 0rem;"
  >
    <button
      class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 invisible"
    >
      <svg
        class="fill-white"
      />
    </button>
    <pre
      class="text-amber-700"
    >
      key
      :
    </pre>
     
    <pre
      class="flex items-center space-x-0.5"
    >
      {
      <button
        class="mx-0.5 rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
      >
        <svg />
      </button>
      }
    </pre>
  </div>
  <div
    class="relative "
  >
    <button
      class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
      style="margin-left: 2rem; width: calc(100% - 2rem); z-index: 2;"
    >
      <div
        class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
      />
      <div
        class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
      >
        <svg
          class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
        />
      </div>
    </button>
    <div
      class="mt-1 flex items-center space-x-1"
      style="margin-left: 1rem;"
    >
      <button
        class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
      >
        <svg
          class="fill-white"
        />
      </button>
      <pre
        class="text-amber-700"
      >
        hello
        :
      </pre>
       
      <div
        class="flex-auto"
      >
        <pre
          class="whitespace-pre-wrap break-all text-sky-400"
        >
          "world"
        </pre>
      </div>
    </div>
    <button
      class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
      style="margin-left: 2rem; width: calc(100% - 2rem); z-index: 2;"
    >
      <div
        class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
      />
      <div
        class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
      >
        <svg
          class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
        />
      </div>
    </button>
    <div
      class="mt-1 flex items-center space-x-1"
      style="margin-left: 1rem;"
    >
      <button
        class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 invisible"
      >
        <svg
          class="fill-white"
        />
      </button>
      <pre
        class="text-amber-700"
      >
        obj
        :
      </pre>
       
      <pre
        class="flex items-center space-x-0.5"
      >
        {
        <button
          class="mx-0.5 rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
        >
          <svg />
        </button>
        }
      </pre>
    </div>
    <div
      class="relative hidden"
    >
      <button
        class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
        style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
      >
        <div
          class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
        />
        <div
          class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
        >
          <svg
            class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
          />
        </div>
      </button>
      <div
        class="mt-1 flex items-center space-x-1"
        style="margin-left: 2rem;"
      >
        <button
          class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
        >
          <svg
            class="fill-white"
          />
        </button>
        <pre
          class="text-amber-700"
        >
          barbes
          :
        </pre>
         
        <div
          class="flex-auto"
        >
          <pre
            class="whitespace-pre-wrap break-all text-sky-400"
          >
            "rochechouart"
          </pre>
        </div>
      </div>
      <button
        class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
        style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
      >
        <div
          class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
        />
        <div
          class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
        >
          <svg
            class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
          />
        </div>
      </button>
      <div
        class="mt-1 flex items-center space-x-1"
        style="margin-left: 2rem;"
      >
        <button
          class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
        >
          <svg
            class="fill-white"
          />
        </button>
        <pre
          class="text-amber-700"
        >
          version
          :
        </pre>
         
        <div
          class="flex-auto"
        >
          <pre
            class="whitespace-pre-wrap break-all text-rose-500"
          >
            1
          </pre>
        </div>
      </div>
      <button
        class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
        style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
      >
        <div
          class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
        />
        <div
          class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
        >
          <svg
            class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
          />
        </div>
      </button>
      <div
        class="mt-1 flex items-center space-x-1"
        style="margin-left: 2rem;"
      >
        <button
          class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
        >
          <svg
            class="fill-white"
          />
        </button>
        <pre
          class="text-amber-700"
        >
          toto
          :
        </pre>
         
        <div
          class="flex-auto"
        >
          <pre
            class="whitespace-pre-wrap break-all text-purple-500"
          >
            null
          </pre>
        </div>
      </div>
      <button
        class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
        style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
      >
        <div
          class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
        />
        <div
          class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
        >
          <svg
            class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
          />
        </div>
      </button>
      <div
        class="mt-1 flex items-center space-x-1"
        style="margin-left: 2rem;"
      >
        <button
          class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 invisible"
        >
          <svg
            class="fill-white"
          />
        </button>
        <pre
          class="text-amber-700"
        >
          arr
          :
        </pre>
         
        <pre
          class="flex items-center space-x-0.5"
        >
          [
          <button
            class="mx-0.5 rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
          >
            <svg />
          </button>
          ]
        </pre>
      </div>
      <div
        class="relative hidden"
      >
        <button
          class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
          style="margin-left: 4rem; width: calc(100% - 4rem); z-index: 4;"
        >
          <div
            class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
          />
          <div
            class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
          >
            <svg
              class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
            />
          </div>
        </button>
        <div
          class="mt-1 flex items-center space-x-1"
          style="margin-left: 3rem;"
        >
          <button
            class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
          >
            <svg
              class="fill-white"
            />
          </button>
          <pre
            class="text-amber-700"
          >
            0
            :
          </pre>
           
          <div
            class="flex-auto"
          >
            <pre
              class="whitespace-pre-wrap break-all text-rose-500"
            >
              0.7309955249202127
            </pre>
          </div>
        </div>
        <button
          class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
          style="margin-left: 4rem; width: calc(100% - 4rem); z-index: 4;"
        >
          <div
            class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
          />
          <div
            class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
          >
            <svg
              class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
            />
          </div>
        </button>
        <div
          class="mt-1 flex items-center space-x-1"
          style="margin-left: 3rem;"
        >
          <button
            class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
          >
            <svg
              class="fill-white"
            />
          </button>
          <pre
            class="text-amber-700"
          >
            1
            :
          </pre>
           
          <div
            class="flex-auto"
          >
            <pre
              class="whitespace-pre-wrap break-all text-rose-500"
            >
              0.45152102178420384
            </pre>
          </div>
        </div>
        <button
          class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
          style="margin-left: 4rem; width: calc(100% - 4rem); z-index: 4;"
        >
          <div
            class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
          />
          <div
            class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
          >
            <svg
              class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
            />
          </div>
        </button>
        <div
          class="mt-1 flex items-center space-x-1"
          style="margin-left: 3rem;"
        >
          <button
            class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
          >
            <svg
              class="fill-white"
            />
          </button>
          <pre
            class="text-amber-700"
          >
            2
            :
          </pre>
           
          <div
            class="flex-auto"
          >
            <pre
              class="whitespace-pre-wrap break-all text-rose-500"
            >
              0.3101491368469109
            </pre>
          </div>
        </div>
        <button
          class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
          style="margin-left: 4rem; width: calc(100% - 4rem); z-index: 4;"
        >
          <div
            class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
          />
          <div
            class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
          >
            <svg
              class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
            />
          </div>
        </button>
        <div
          class="mt-1 flex items-center space-x-1"
          style="margin-left: 3rem;"
        >
          <button
            class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
          >
            <svg
              class="fill-white"
            />
          </button>
          <pre
            class="text-amber-700"
          >
            3
            :
          </pre>
           
          <div
            class="flex-auto"
          >
            <pre
              class="whitespace-pre-wrap break-all text-rose-500"
            >
              0.33964078594296154
            </pre>
          </div>
        </div>
        <button
          class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
          style="margin-left: 4rem; width: calc(100% - 4rem); z-index: 4;"
        >
          <div
            class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
          />
          <div
            class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
          >
            <svg
              class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
            />
          </div>
        </button>
        <div
          class="mt-1 flex items-center space-x-1"
          style="margin-left: 3rem;"
        >
          <button
            class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
          >
            <svg
              class="fill-white"
            />
          </button>
          <pre
            class="text-amber-700"
          >
            4
            :
          </pre>
           
          <div
            class="flex-auto"
          >
            <pre
              class="whitespace-pre-wrap break-all text-rose-500"
            >
              0.579201406715314
            </pre>
          </div>
        </div>
        <button
          class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
          style="margin-left: 4rem; width: calc(100% - 4rem); z-index: 4;"
        >
          <div
            class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
          />
          <div
            class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
          >
            <svg
              class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
            />
          </div>
        </button>
      </div>
      <button
        class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
        style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
      >
        <div
          class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
        />
        <div
          class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
        >
          <svg
            class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
          />
        </div>
      </button>
      <div
        class="mt-1 flex items-center space-x-1"
        style="margin-left: 2rem;"
      >
        <button
          class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 invisible"
        >
          <svg
            class="fill-white"
          />
        </button>
        <pre
          class="text-amber-700"
        >
          nested
          :
        </pre>
         
        <pre
          class="flex items-center space-x-0.5"
        >
          {
          <button
            class="mx-0.5 rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
          >
            <svg />
          </button>
          }
        </pre>
      </div>
      <div
        class="relative hidden"
      >
        <button
          class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
          style="margin-left: 4rem; width: calc(100% - 4rem); z-index: 4;"
        >
          <div
            class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
          />
          <div
            class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
          >
            <svg
              class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
            />
          </div>
        </button>
        <div
          class="mt-1 flex items-center space-x-1"
          style="margin-left: 3rem;"
        >
          <button
            class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
          >
            <svg
              class="fill-white"
            />
          </button>
          <pre
            class="text-amber-700"
          >
            thx
            :
          </pre>
           
          <div
            class="flex-auto"
          >
            <pre
              class="whitespace-pre-wrap break-all text-rose-500"
            >
              1138
            </pre>
          </div>
        </div>
        <button
          class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
          style="margin-left: 4rem; width: calc(100% - 4rem); z-index: 4;"
        >
          <div
            class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
          />
          <div
            class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
          >
            <svg
              class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
            />
          </div>
        </button>
        <div
          class="mt-1 flex items-center space-x-1"
          style="margin-left: 3rem;"
        >
          <button
            class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
          >
            <svg
              class="fill-white"
            />
          </button>
          <pre
            class="text-amber-700"
          >
            star
            :
          </pre>
           
          <div
            class="flex-auto"
          >
            <pre
              class="whitespace-pre-wrap break-all text-sky-400"
            >
              "wars"
            </pre>
          </div>
        </div>
        <button
          class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
          style="margin-left: 4rem; width: calc(100% - 4rem); z-index: 4;"
        >
          <div
            class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
          />
          <div
            class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
          >
            <svg
              class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
            />
          </div>
        </button>
      </div>
      <button
        class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
        style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
      >
        <div
          class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
        />
        <div
          class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
        >
          <svg
            class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
          />
        </div>
      </button>
    </div>
    <button
      class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
      style="margin-left: 2rem; width: calc(100% - 2rem); z-index: 2;"
    >
      <div
        class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
      />
      <div
        class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
      >
        <svg
          class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
        />
      </div>
    </button>
    <div
      class="mt-1 flex items-center space-x-1"
      style="margin-left: 1rem;"
    >
      <button
        class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
      >
        <svg
          class="fill-white"
        />
      </button>
      <pre
        class="text-amber-700"
      >
        done
        :
      </pre>
       
      <div
        class="flex-auto"
      >
        <pre
          class="whitespace-pre-wrap break-all text-green-500"
        >
          true
        </pre>
      </div>
    </div>
    <button
      class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
      style="margin-left: 2rem; width: calc(100% - 2rem); z-index: 2;"
    >
      <div
        class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
      />
      <div
        class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
      >
        <svg
          class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
        />
      </div>
    </button>
  </div>
</div>
`;
